File: app/src/main/java/com/example/development/sakaiclient20/networking/deserializers/builders/courses/CourseBuilder.java
package com.example.development.sakaiclient20.networking.deserializers.builders.courses;

import com.example.development.sakaiclient20.models.Term;
import com.example.development.sakaiclient20.networking.deserializers.builders.AbstractBuilder;
import com.example.development.sakaiclient20.persistence.entities.Course;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;

/**
 * Created by Development on 8/5/18.
 */

public class CourseBuilder extends AbstractBuilder<JsonObject, Course> {

    public CourseBuilder(JsonObject jsonObject) {
        super(jsonObject);
    }

    @Override
    public CourseBuilder build() {
        // Initialize the course with the siteId
        result = new Course(source.get("id").getAsString());

        result.term = parseTerm();
        result.subjectCode = parseSubjectCode();
        result.title = source.get("title").getAsString();

        JsonElement desc = source.get("description");

        result.description = !desc.isJsonNull()
                ? desc.getAsString()
                : "";
        result.siteOwner = source.getAsJsonObject("siteOwner")
                .get("userDisplayName")
                .getAsString();

        JsonArray rawSitePages = source.getAsJsonArray("sitePages");
        SitePagesBuilder builder = new SitePagesBuilder(rawSitePages);
        result.sitePages = builder.build().getResult();
        result.assignmentSitePageUrl = builder.getAssignmentSitePageUrl();

        return this;
    }

    private Term parseTerm() {
        JsonObject props = source.getAsJsonObject("props");
        JsonElement termEidElement = props.get("term_eid");

        // It is possible that even if props is defined, term_eid
        // is not provided, in which case the element itself will
        // be null (instead of holding a null value and being JsonNull)
        if(termEidElement == null)
            return new Term("0000:0");

        String termEid = termEidElement.getAsString();
        return new Term(termEid);
    }

    private int parseSubjectCode() {
        JsonElement providerGroupIdElement = source.get("providerGroupId");
        if(providerGroupIdElement instanceof JsonNull)
            return 0;

        String providerGroupId = providerGroupIdElement.getAsString();
        providerGroupId = providerGroupId.replace("+", "_delim_");
        String courseCode = providerGroupId.split("_delim_")[0];
        String subjectCode = courseCode.split(":")[3];
        return Integer.parseInt(subjectCode);
    }
}

File: app/src/main/java/com/sakaimobile/development/sakaiclient20/networking/deserializers/builders/courses/CourseBuilder.kt
package com.sakaimobile.development.sakaiclient20.networking.deserializers.builders.courses

import com.google.gson.JsonObject
import com.sakaimobile.development.sakaiclient20.models.Term
import com.sakaimobile.development.sakaiclient20.networking.deserializers.builders.AbstractBuilder
import com.sakaimobile.development.sakaiclient20.networking.deserializers.getStringMember
import com.sakaimobile.development.sakaiclient20.persistence.entities.Course

/**
 * Created by Development on 8/5/18.
 */

class CourseBuilder(jsonObject: JsonObject) : AbstractBuilder<JsonObject, Course>(jsonObject) {

    override fun build(): CourseBuilder {
        // Initialize the course with the siteId
        result = Course(source.getStringMember("id")!!)

        result.term = parseTerm()
        result.subjectCode = parseSubjectCode()
        result.title = source.getStringMember("title")

        result.description = source.getStringMember("description")
        result.siteOwner = source.getAsJsonObject("siteOwner")
                                 .getStringMember("userDisplayName")

        val rawSitePages = source.getAsJsonArray("sitePages")
        val builder = SitePagesBuilder(rawSitePages)
        result.sitePages = builder.build().result
        result.assignmentSitePageUrl = builder.getAssignmentSitePageUrl()

        return this
    }

    private fun parseTerm(): Term {
        // It is possible that even if props is defined, term_eid
        // is not provided, in which case the element itself will
        // be null (instead of holding a null value and being JsonNull)
        val props = source.getAsJsonObject("props")
        val termEid = props.getStringMember("term_eid", default = "0000:0")
        return Term(termEid)
    }

    private fun parseSubjectCode(): Int {
        val providerGroupId = source.getStringMember("providerGroupId", default = null)
                                    ?.replace("+", "_delim_")
                                    ?: return 0
        // dropLastWhile removes trailing empty strings, just good practice when splitting
        val courseCode = providerGroupId.split("_delim_".toRegex())
                                        .dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        val subjectCode = courseCode.split(":".toRegex())
                                    .dropLastWhile { it.isEmpty() }.toTypedArray()[3]
        return Integer.parseInt(subjectCode)
    }
}
